clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -2458,6 +2458,11 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
       &getTarget().getLongDoubleFormat() == &llvm::APFloat::IEEEquad())
     BuiltinID = mutateLongDoubleBuiltin(BuiltinID);
 
+   // Mutate the printf builtin ID so that we use the same CodeGen path for
+   // HIP and OpenCL with AMDGPU targets.
+   if (getTarget().getTriple().isAMDGCN() && BuiltinID == AMDGPU::BIprintf)
+     BuiltinID = Builtin::BIprintf;
```

Comment:
I'm very close to landing 'real' printf support in the GPU libc where `printf` is just a regular function call. Will this change the handling for that in any way? I've already had to make the backend pass respect `-fno-builtins` and remove `ockl` from OpenMP to make that possible so I'm hoping we don't end up with a lot more special casing for `printf`.

Reply:
@jhuber6 I do not believe there are any current plans to use GPU libc from HIP or OpenCL.  So there will continue to be a division between OpenMP and HIP and OpenCL printf handling.

Reply:
If we do the eager replacement of `printf` that HIP and OpenCL uses currently then it won't be linked in. So users should still be able to link in stuff like `strcmp` or whatever without it interfering. This would require the new driver however, and if they attempted to use something like `fputs` it would segfault because no one initialized the buffer, which isn't a terrible failure mode all things considered.

Reply:
@jhuber6 , I had your implementation in mind when I wrote this, The printf wont be expanded by clang with "-fno-builtin" and users would still have an option to use a lib variant if need be.  This also makes this more elegant as we would not have to hack the "fno-builtin" handling into the implementation, this is part of clang builtin handling.